### ISSUE TITLE
C++ function returning std::function with reference argument is not recognized as a function (Origin: bugzilla #792830)

### ISSUE THREAD

**Issue by doxygen** [doxygen/doxygen#6319]
status NEW severity *normal* in component *general* for *---*
Reported in version *1.8.14* on platform *Other*
Assigned to: Dimitri van Heesch

On 2018-01-23 15:44:39 +0000, jarle.code@gmail.com wrote:

    For a function returning a std::function object, if there is a reference in
    the std::function template argument Doxygen erronously classifies the
    function as a variable.
    
    Example:
    -----------------------------------------------
    #include <functional>
    
    /** Test function */
    std::function<void(const float&)> MyTestFunc();
    -----------------------------------------------
    MyTestFunc is listed as a variable in the generated html documentation.
    
    The following code is parsed correctly:
    -----------------------------------------------
    #include <functional>
    
    /** Test function */
    std::function<void(const float)> MyTestFunc();
    -----------------------------------------------

**Comment by albert-github** [comment]
Looks like that in `scanner.l` in the rule:
```
<Function>[;{,]
```
the regular expression:
```
static QRegExp re("([^)]*[*&][^)]*)"); // (...*...)
```
and consequently:
```
                                             if (!current->type.isEmpty() &&
                                                 (current->type.find(re,0)!=-1 || current->type.left(8)=="typedef "))
```
is causing the problems.

Case: [example.tar.gz](https://github.com/doxygen/doxygen/files/3882409/example.tar.gz)

**Comment by Sygmei** [comment]
I'm having the same problem, Doxygen is detecting the following signature
```cpp
std::function<void(kaguya::State*)> InitTreeNodeAsTable(const std::string& nodeName);
```
as a variable.

What is the status of this bug ?

**Comment by doxygen** [comment]
@Sygmei I've just pushed a possible solution. Let me know if that fixes the problem for you. 
Please don't close this bug report yet, this will be done when a new official release has been created.

**Comment by Sygmei** [comment]
@doxygen it fixed the problem :) Thanks a lot !

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.18.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
